### DISCUSSION TITLE
emoji重名覆盖

### DISCUSSION THREAD

**Discussion by lrhtony** [walinejs/waline#1214] [Ideas]
https://github.com/walinejs/waline/blob/08a5c60c92277aea344002757dd1cc3cf97bde1d/packages/client/src/utils/emoji.ts#L62-L67
自行测试+看代码，如果不同emoji中存在重名，key相同，前面的表情url就会被后面相同key的表情所覆盖

个人想法是可以参照Discord对重名表情的处理，在已经存在key的情况下，给后面相同的key加上“~1” “~2“

如"test"， 后面的就为“test\~1”“test\~2” ...
自己稍微改了一下https://github.com/lrhtony/waline/commit/414616abe71befa0e53baf4dda30678a59345369

**Comment by lizheming** [comment]
为啥会有 key 重复的场景？选择表情那可以这么处理，那评论里怎么处理呢，同一个 key 我怎么知道是要渲染 A 还是 B 呢？

**Reply by lrhtony** [reply]
渲染的话是按链接来渲染
我这种情况指的是https://cdn.jsdelivr.net/gh/walinejs/emojis@1.0.0/bilibili，https://cdn.jsdelivr.net/gh/walinejs/emojis@1.0.0/tieba 类似这种里面info.json各自独立，那么两者分别拼接后的表情key就可能存在同key的情况

发表评论时的评论渲染也是查询emojiConfig，我这样是直接把原来相同的key变成不同key，发表时就能分清渲染A还是B

至于评论列表渲染，因为数据存储时存的是链接，原始链接不同，自然就不受影响

**Reply by lizheming** [reply]
1. 评论列表渲染的时候存储的是 key 不是链接，所以会受到影响。使用一个 hash map 的原因也正是想反响映射的时候更方便的替换表情
2. 不能依赖顺序动态的改变 key，这是因为顺序会随着作者的配置调整而调整，当其在之前插入了某个带同名 key 的表情包渲染就会不符合预期

**Comment by Mister-Hope** [comment]
我不能理解这个需求，作为一个访客想要评论，我为什么会知道你的emoji是什么key值？换句话说，除了markdown-it-emoji规定的公认的emoji名称，我怎么知道我输入什么key是有效的？

访客本来就应该通过表情选择器输入表情，而且本身waline已经支持markdown-it-emoji的所有标准emoji名称，这已经可以覆盖大多数用户想要“自己敲一个emoji”的场景了。

我想不到你非得要把key的prefix去掉，除了带来一堆负面效果的同时让你作为熟悉emoji名称的网站主稍微评论的舒心之外，对访客有什么裨益
